### DISCUSSION TITLE
Plugin universal_metadata (same metadata structure for iOS & Android)

### DISCUSSION THREAD

**Discussion by LaGregance** [fastlane/fastlane#21027] [Show and tell]
Hi, I just finish to publish a plugin that is really helpful for managing metadata for iOS & Android in the same project (useful with React Native).

You can find the project here : https://github.com/LaGregance/fastlane-plugin-universal_metadata

# Genesis
By default on Fastlane we use `upload_to_app_store` & `upload_to_play_store` to publish metadata on the store. But each of them use different folder structure to achieve the same goal.

The goal of this Fastlane plugin is to unify metadata & screenshots structure for both iOS and Android, and finally produce metadata folder that will be used by `upload_to_app_store` & `upload_to_play_store`.

# Features

- Generate every screenshot size derived from a single file
- Generate every metadata for both Android & iOS from a single json file

# Setup

1. First install the plugin by using `fastlane add_plugin universal_metadata`
2. Run the command `fastlane run universal_metdata init:true languages:"en-US,fr-FR"`, this will init the folder `universal-metadata` with this structure
    
  ```
  universal-metadata
  ├── screenshots
  │   └── default
  │   └── en-US
  │   └── fr-FR
  ├── description
  │   └── default
  │       ├── full_description.txt
  │       └── short_description.txt
  │   └── en-US
  │       └── ...
  │   └── fr-FR
  │       └── ...
  ├── release-notes
  │   ├── default.txt
  │   ├── en-US.txt
  │   └── fr-FR.txt
  └── metadata.json
  ```
    
## Result

After running `fastlane run universal_metdata languages:"en-US,fr-FR"` you will get the following result:
```
screenshots
│   └── fr
│       ├── screen-1.png
│       └── screen-2.png
metadata
├── android
│   └── fr
│       ├── images
│       │   ├── phoneScreenshots
│       │   │   ├── screen-1.png
│       │   │   └── screen-2.png
│       │   ├── sevenInchScreenshots
│       │   │   ├── screen-1.png
│       │   │   └── screen-2.png
│       │   └── tenInchScreenshots
│       │   │   ├── screen-1.png
│       │   │   └── screen-2.png
│       ├── full_description.txt
│       ├── short_description.txt
│       ├── title.txt
│       └── video.txt
├── fr
│   ├── name.txt
│   ├── subtitle.txt
│   ├── privacy_url.txt
│   ├── description.txt
│   ├── keywords.txt
│   ├── release_notes.txt
│   ├── support_url.txt
│   ├── marketing_url.txt
│   └── promotional_text.txt
├── copyright.txt
├── primary_category.txt
├── secondary_category.txt
├── primary_first_sub_category.txt
├── primary_second_sub_category.txt
├── secondary_first_sub_category.txt
├── secondary_second_sub_category.txt
├── first_name.txt
├── last_name.txt
├── phone_number.txt
├── email_address.txt
├── demo_user.txt
├── demo_password.txt
└── notes.txt
```
Now you are able to run `upload_to_app_store` & `upload_to_play_store` without pain.

# Repo
More infos: https://github.com/LaGregance/fastlane-plugin-universal_metadata

**Comment by slavdok** [comment]
Oh this is so great. Will it fetch existing metadata from the respective stores as well? Will it conflict if I try to fetch into this structure?

**Reply by LaGregance** [reply]
Unfortunately it will not fetch metadata from the store, also if you fetch metadata yourself, you will have to manually move them to the `universal-metadata` folder because the `metadata` folder is re-created each time you run the command.
Indeed it should be great to have add a command to synchronise actual metadata from the store and integrate them into the universal-metadata structure.
I'll let you know when I have time to look into it
